### PR TITLE
Async workshop

### PR DESCRIPTION
**Pull request by diondokter**

### CODE REVIEW

File: slides/C-advanced-rust.md [hunk 1]
```diff
@@ -618,6 +618,367 @@ layout: default
...
+- Introduction Rust `async` programming
+- Uncovering `Future` type
+- Mechanics behind `async`/`await`
+- Creating our own async primitives
+
+---
+layout: default
+---
+
+# What is async rust?
```
Comment (hdoordt) [inline slides/C-advanced-rust.md:637]:
```suggestion
# What is async Rust?
```

File: slides/C-advanced-rust.md [hunk 2]
```diff
@@ -618,6 +618,367 @@ layout: default
...
+
+---
+layout: two-cols
+---
+
+# Async block?
+
+Rust syntax
+
+Instructs the compiler to turn the code block into a statemachine
```
Comment (hdoordt) [inline slides/C-advanced-rust.md:676]:
```suggestion
Instructs the compiler to turn the code block into a state machine
```

File: slides/C-advanced-rust.md [hunk 3]
```diff
@@ -618,6 +618,367 @@ layout: default
...
+    let c = (a + b + rand()).to_string();
+    sleep().await;
+    c
+};
+```
+
+::right::
+
+Only data carried over await points are kept.
+This statemachine would roughly look like:
```
Comment (hdoordt) [inline slides/C-advanced-rust.md:693]:
```suggestion
This state machine would roughly look like:
```

File: slides/C-advanced-rust.md [hunk 4]
```diff
@@ -618,6 +618,367 @@ layout: default
...
+        a: u32,
+        b: u32,
+    }
+    Stage2 {
+        c: String,
+    }
+}
+```
+
+But how do we interact with the statemachine?
```
Comment (hdoordt) [inline slides/C-advanced-rust.md:710]:
```suggestion
But how do we interact with the state machine?
```

File: slides/C-advanced-rust.md [hunk 5]
```diff
@@ -618,6 +618,367 @@ layout: default
...
+
+```rust
+pub trait Future {
+    type Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
+}
+```
+
+- Async statemachines implement `Future`
```
Comment (hdoordt) [inline slides/C-advanced-rust.md:728]:
```suggestion
- Async state machines implement `Future`
```

File: slides/C-advanced-rust.md [hunk 6]
```diff
@@ -618,6 +618,367 @@ layout: default
...
+
+---
+layout: two-cols
+---
+
+# Polling a future
+
+- The thing doing the polling is called the executor
+- Continuously polling is inefficient
+  - Cpu needs to stay active
```
Comment (hdoordt) [inline slides/C-advanced-rust.md:751]:
```suggestion
  - CPU needs to stay active
```

**Review by hdoordt — CHANGES_REQUESTED** [review]
Nice! Some small nits which I'll change

**Comment by diondokter** [comment]
Done

**Review by hdoordt — APPROVED** [review]

**Pull request merged**